mne_bids/read.py

```diff
@@ -71,6 +72,89 @@ def _read_raw(raw_fname, electrode=None, hsp=None, hpi=None, config=None,
     return raw
 
 
+def _handle_events_reading(events_fname, raw):
+    """Read associated events.tsv and populate raw.
+
+    Handle onset, duration, and description of each event.
+    """
+    events_dict = _from_tsv(events_fname)
+
+    # Get the descriptions of the events
+    if 'trial_type' in events_dict:
+        descriptions = np.asarray(events_dict['trial_type'], dtype=str)
+        # Drop events unrelated to a trial type
+        events_dict = _drop(events_dict, 'n/a', 'trial_type')
+
+    # If we don't have a proper description of the events, perhaps we have
+    # at least an event value?
+    elif 'value' in events_dict:
+        descriptions = np.asarray(events_dict['value'], dtype=str)
+        # Drop events unrelated to value
+        events_dict = _drop(events_dict, 'n/a', 'value')
+
+    # Worst case, we go with 'n/a' for all events
+    else:
+        descriptions = 'n/a'
+
+    # Add Events to raw as annotations
+    onsets = np.asarray(events_dict['onset'], dtype=float)
+    durations = np.asarray(events_dict['duration'], dtype=float)
+    annot_from_events = mne.Annotations(onset=onsets,
+                                        duration=durations,
+                                        description=descriptions,
+                                        orig_time=raw.info['meas_date'])
+    raw.set_annotations(annot_from_events)
+
+    return raw
+
+
+def _handle_channels_reading(channels_fname, raw):
+    """Read associated channels.tsv and populate raw.
+
+    Updates status (bad) and types of channels.
+    """
+    channels_dict = _from_tsv(channels_fname)
+
+    # If we have a channels.tsv file, make sure there is the optional "status"
+    # column from which to infer good and bad channels
+    if 'status' in channels_dict:
+        # find bads from channels.tsv
+        bad_bool = [True if chn == 'bad' else False
+                    for chn in channels_dict['status']]
+        bads = np.asarray(channels_dict['name'])[bad_bool]
+
+        # merge with bads already present in raw data file (if there are any)
+        unique_bads = set(raw.info['bads']).union(set(bads))
+        raw.info['bads'] = list(unique_bads)
+
+    # Furthermore, try to update the channel types according to the "type"
+    # and "name" columns in channels.tsv. These are mandatory in BIDS
+    # First, make sure that ordering of names in channels.tsv matches the
+    # ordering of names in the raw data.
```

Comment:
This seems a bit fragile. Why don't you just find the correct channel and update it?

Reply:
because we cannot be certain that the `name` column of the JSON corresponds exactly to the names in `raw.ch_names`. In BIDS, the only prescription is that the **order** of channels SHOULD be the same between the channels.tsv and the raw data file

Reply:
those are really loose rules, and in my opinion, they could be tightened a bit:
- entries in channels.tsv MUST be **equal** to those in the raw data file
- order of channels MUST be the same

Reply:
> he order of channels SHOULD be the same between the channels.tsv and the raw data file

note this is different from [what is in the specification](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/03-electroencephalography.html#channels-description-_channelstsv). There is no capitalization there so should loses the special meaning IMO

It's fine if BIDS does not impose the entries to be equal. As a software, we can impose that. I fear you would have surprising results if you do it like this.

Reply:
> There is no capitalization there so should loses the special meaning IMO

So far I did not think that capitalization changes the meaning of "should" ... unfortunately, there is nothing on this in the [spec](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html) or the [RFC](https://www.ietf.org/rfc/rfc2119.txt), where it just says vaguely:

> These words are often capitalized.


But anyhow, in this case I am pretty sure that it's a typo and the intent was the meaning of SHOULD as in the RFC.

Reply:
But fine with me to impose this from the side of MNE-BIDS and raise an error if we find a different situation :-)